efi/env.go

```diff
@@ -0,0 +1,305 @@
+}
+
+type varReadWriter interface {
+	varReader
+
+	// WriteVar updates the specified EFI variable.
+	WriteVar(name string, guid efi.GUID, attrs efi.VariableAttributes, data []byte) error
+}
+
+// varBranchState corresponds to a EFI variable state associated with a profile branch,
+// consisting of an initial starting environment and a sequence of updates that have been
+// added during profile generation. Each profile generation loop starts with one of these
+// without any updates. Branches in a profile inherit a copy of this from the parent branch
+// and may make modifications to EFI variables (eg, applying a SBAT update) which may affect
+// other branches - in this case, the profile generation may be re-executed multiple times
+// with different starting states as a result of these updates.
+type varBranchState struct {
+	initial   *rootVarReader // the initial starting environment
+	updates   *varUpdate     // the updates applied by the associated branch
+	collector *rootVarsCollector
+}
+
+func newVarBranchState(initial *rootVarReader, collector *rootVarsCollector) *varBranchState {
+	return &varBranchState{initial: initial, collector: collector}
+}
+
+// ReadVar implements varReader.ReadVar.
+func (s *varBranchState) ReadVar(name string, guid efi.GUID) ([]byte, efi.VariableAttributes, error) {
+	update := s.updates
+	for ; update != nil; update = update.previous {
+		desc := efi.VariableDescriptor{Name: name, GUID: guid}
+		if update.name == desc {
+			return update.data, update.attrs, nil
+		}
+	}
+	return s.initial.ReadVar(name, guid)
+}
+
+// WriteVar implements varReadWriter.WriteVar.
+func (s *varBranchState) WriteVar(name string, guid efi.GUID, attrs efi.VariableAttributes, data []byte) error {
+	orig, _, err := s.ReadVar(name, guid)
+	switch {
+	case err == efi.ErrVarNotExist:
+		// ok
+	case err != nil:
+		return err
+	}
+
+	if attrs&efi.AttributeAppendWrite != 0 {
+		data = append(orig, data...)
+	}
+
+	s.updates = &varUpdate{
+		previous: s.updates,
+		name:     efi.VariableDescriptor{Name: name, GUID: guid},
+		attrs:    attrs &^ efi.AttributeAppendWrite,
+		data:     data}
+	s.collector.registerUpdatesFor(s.initial, s.updates)
+	return nil
+}
+
+// rootVarsCollector keeps track of all of the starting EFI variable states that
+// a profile needs to be generated against. The profile generation runs an outer
+// loop that adds a branch for each of the starting states. Profile generation may
+// add more starting states as some branches have paths that update EFI variables
+// (such as applying SBAT updates) which may have an effect on other branches.
+type rootVarsCollector struct {
+	host           varReader                       // the externally supplied host environment
+	pendingUpdates map[*rootVarReader][]*varUpdate // update sequences that haven't been processed yet
+	seen           map[rootVarReaderKey]struct{}   // seen starting states
+	todo           []*rootVarReader                // starting states still to process
```

Comment:
pendingUpdates and todo probably need a bit more commentary here, as otherwise is fairly unclear what is their relationship. It seems todos contain starting states where update sequence have been applied?

Reply:
I've tried to clarify this a bit more now.